Callback support and learn packet enhancement

AMairesse opened the pull request:
Add support for a callback function
Add reading of 'func' and 'type' when receiving a learn packet

kipe commented on enocean/communicators/communicator.py:
```diff
@@ -58,7 +60,10 @@ def parse(self):
+            # If message is OK, add it to receive queue or send to the callback method
             if status == PARSE_RESULT.OK and p:
-                self.receive.put(p)
+                if self.__callback == None:
```
Comparison should be `if self.__callback is None:` according to PEP8

kipe commented on enocean/protocol/packet.py:
```diff
@@ -169,6 +171,12 @@ def parse(self):
                 self.contains_eep = self.bit_data[-7]
+        
+        # Get rorg_func and rorg_type from a learn packet
+        if (self.learn):
```
Remove unnecessary quotes: `if self.learn:`

kipe commented:
Thanks for your commit :smile: 
I added some comments, PEP8-fixes. If you could fix those, it would be great.

Can you also point me to the documentation of this feature, as seems like I've missed it :confused:

AMairesse commented:
Hi,

Ok for the fixes, I'll update now, thanks for your review.

About the documentation I don't understand your question or see a file I could append.
* "callback" support is just an option for a client application to directly use it's own function. Useful because the library is already threaded (I was inspired from the way this one works : https://pypi.python.org/pypi/XBee)
* the "func" and "type" are the parameters used to interpret the EEP. They are inside the packet when the learn bit is 'True'. It's useful for a client application to be able to "register" different devices.

Antoine.

kipe commented:
Hey,

I was referring to the `func` and `type` parameters. I'd like to see some documentation related to this, as I have missed this feature completely. I just tried with the sensors I have, and at least they don't the information. I also checked the documentation and some EEP-profiles do use those bits for other purposes, so the function isn't valid at the moment (invalid values may be put to `self.rorg_func` and `self.rorg_type`).

I think this feature is specific to `self.contains_eep` and should thus be handled only if it's set (around line 170 in packet.py). Although, looking at the `EnOcean_Equipment_Profiles_EEP_V2.61_public.pdf` -documentation I'm not sure if the correct bit is `-8` or `-7` in `self.contains_eep = self.bit_data[-7]`.

If you could check the output of both bits (`self.bit_data[-7]` and `self.bit_data[-8]`), it would be great. If you can't, can you please post the whole message output from the debug-log and I can try to figure this out. Also, it would be helpful to know what EEP-profile (`self.rorg`, `self.rorg_func` and `self.rorg_type`) you're getting and which sensor are you using.

AMairesse commented:
You were right, the correct bit for contains_eep should be -8 which is "LRN Type". The -7 is the "EEP result" which is used on bidirectional learn process.

The documentation I found is here : http://www.enocean-alliance.org/eep/ (2.6.3 as of now)
On pages 182/183 the EEP profile exchange is explained. There is 3 variations : 
* variation 1 : cannot have the EEP profile. This is a teach-in telegram but LRN Type have the 0 value
* variation 2 : you have the EEP profile in the telegram. Same as variation 1 but LRN Type is set to 1
* variation 3 : there is a bidirectional communication. I didn't dig a lot but I think it wouldn't work and would need a rework. Adding a way for the library to manage this would require to save the data... I don't have such a component to explore this.

Here is the log from my sensor : 
01:8A:7B:30->FF:FF:FF:FF (-73 dBm): 0x01 ['0xa5', '0x8', '0x28', '0x46', '0x80', '0x1', '0x8a', '0x7b', '0x30', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x49', '0x0'] {}
2015-08-10 18:32:04,271 - enocean.communicators.Communicator - DEBUG - 01:8A:7B:30->FF:FF:FF:FF (-73 dBm): 0x01 ['0xa5', '0x8', '0x28', '0x46', '0x80', '0x1', '0x8a', '0x7b', '0x30', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x49', '0x0'] {}

When my application get the rorg_func and rorg_type data from this first telegram I can decode the EEP, which give : 
2015-08-10 18:32:08,691 - enocean.communicators.Communicator - DEBUG - 01:8A:7B:30->FF:FF:FF:FF (-76 dBm): 0x01 ['0xa5', '0x8', '0x28', '0x46', '0x80', '0x1', '0x8a', '0x7b', '0x30', '0x0'] ['0x1', '0xff', '0xff', '0xff', '0xff', '0x4c', '0x0'] {u'TMP': {u'value': 29.019607843137255, u'description': u'Temperature (linear)', u'unit': u'C', u'raw_value': 70}}

This is a temperature sensor and the rorg, func and type I get are correct, same as the values written on the sensor : 
* UBID1504
* EEP: A5-02-05

I've updated my patch : 
* changed the contains_eep to use the -8 bit
* retrieve rorg_func and rorg_type only if this is a 4BS telegram, LRN mode is on and LRN type is 1
Should be ok now for your devices, probably won't get this values so often but should always be right when having it.

Antoine.

kipe commented:
Great, amazing work :+1: 
Merging this now. If you have time and interest, tests for this feature would be great also, so it won't break when making further modifications. I can also make them based on the information you just gave me, once I have the time.

kipe commented:
I created the test in 35ecba1ea2c2d5f0703a348196f6a3aad2f561f5, as I wanted to make some other modifications (not directly related to this, but inspired by it).
